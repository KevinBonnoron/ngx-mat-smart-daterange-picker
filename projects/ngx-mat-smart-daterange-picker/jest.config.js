require('jest-preset-angular/ngcc-jest-processor');

module.exports = {
  preset: 'jest-preset-angular',
  transformIgnorePatterns: ['node_modules/(?!.*\\.mjs$)'],
  snapshotSerializers: [
    'jest-preset-angular/build/serializers/html-comment',
    'jest-preset-angular/build/serializers/ng-snapshot',
    'jest-preset-angular/build/serializers/no-ng-attributes',
  ],
};
