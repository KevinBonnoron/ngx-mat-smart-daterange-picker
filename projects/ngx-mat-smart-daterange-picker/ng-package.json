{
  "$schema": "../../node_modules/ng-packagr/ng-package.schema.json",
  "dest": "../../dist/ngx-mat-smart-daterange-picker",
  "lib": {
    "entryFile": "src/public-api.ts"
  }
}