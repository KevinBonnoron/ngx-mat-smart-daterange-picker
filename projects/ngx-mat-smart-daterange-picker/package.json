{
  "name": "ngx-mat-smart-daterange-picker",
  "description": "Angular Material custom daterange picker",
  "version": "0.0.1",
  "publishConfig": {},
  "repository": {
    "type": "git",
    "url": "ssh://git@github.com/KevinBonnoron/ngx-mat-smart-daterange-picker"
  },
  "peerDependencies": {
    "@angular/cdk": "^13.0.0",
    "@angular/common": "^13.0.0",
    "@angular/core": "^13.0.0",
    "@angular/material": "^13.0.0",
    "ngx-mat-popover": "0.0.7"
  },
  "dependencies": {
    "tslib": "^2.3.0"
  }
}