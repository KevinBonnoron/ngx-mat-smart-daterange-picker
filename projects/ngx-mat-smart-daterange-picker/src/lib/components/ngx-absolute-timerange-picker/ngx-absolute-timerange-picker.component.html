<div>
  <h3>{{ 'absoluteTimeRange' | translate }}</h3>

  <form [formGroup]="formGroup" autocomplete="off">
    <div fxLayout="column">
      <mat-form-field>
        <mat-label>{{ 'from' | translate }}</mat-label>
        <input matInput formControlName="from" (focus)="pickerVisible = true" />
      </mat-form-field>
      <mat-form-field>
        <mat-label>{{ 'to' | translate }}</mat-label>
        <input matInput formControlName="to" (focus)="pickerVisible = true" />
      </mat-form-field>
      <button
        mat-raised-button
        color="primary"
        [disabled]="!formGroup.valid"
        (click)="applyTimeRange(this.formGroup.value)"
      >
        {{ 'applyTimeRange' | translate }}
      </button>
    </div>
  </form>
</div>
