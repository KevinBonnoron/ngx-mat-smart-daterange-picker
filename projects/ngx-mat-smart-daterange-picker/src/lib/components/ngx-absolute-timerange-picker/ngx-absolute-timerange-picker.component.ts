import { formatDate } from '@angular/common';
import { Component, EventEmitter, Input, LOCALE_ID, OnChanges, Output, SimpleChanges, inject } from '@angular/core';
import { FormBuilder, ValidatorFn, Validators } from '@angular/forms';
import { DateRange } from '../../types';
import { RELATIVE_TIME_REGEX, convertStringToDate } from '../../utils';

export interface AbsoluteTimeRange {
  from: string;
  to: string;
}

@Component({
  selector: 'ngx-absolute-timerange-picker',
  templateUrl: './ngx-absolute-timerange-picker.component.html'
})
export class NgxAbsoluteTimeRangePickerComponent implements OnChanges {
  private readonly formBuilder = inject(FormBuilder);
  private readonly locale = inject(LOCALE_ID);

  @Input()
  value!: DateRange;

  @Input()
  min?: Date;

  @Input()
  max?: Date;

  @Output()
  readonly valueChanged = new EventEmitter<DateRange>();

  readonly formGroup = this.formBuilder.nonNullable.group({
    from: ['', [Validators.required, periodValidator()]],
    to: ['', [Validators.required, periodValidator()]]
  });

  pickerVisible = true;

  ngOnChanges(changes: SimpleChanges): void {
    if (changes['value']) {
      const { from, to } = changes['value'].currentValue ?? {};

      if (from) {
        this.formGroup.patchValue(
          {
            from: formatDate(from, 'yyyy-MM-dd HH:mm:ss', this.locale),
            to: formatDate(to, 'yyyy-MM-dd HH:mm:ss', this.locale)
          },
          { emitEvent: false }
        );
      }
    }

    if (changes['min'] || changes['max']) {
      this.formGroup
        .get('from')
        ?.setValidators(Validators.compose([Validators.required, periodValidator({ min: this.min, max: this.max })]));
      this.formGroup
        .get('to')
        ?.setValidators(Validators.compose([Validators.required, periodValidator({ min: this.min, max: this.max })]));
    }
  }

  applyTimeRange(absoluteTimeRange: Partial<AbsoluteTimeRange>): void {
    if (!absoluteTimeRange.from || !absoluteTimeRange.to) {
      return;
    }

    this.valueChanged.emit({
      from: convertStringToDate(absoluteTimeRange.from, 'from'),
      to: convertStringToDate(absoluteTimeRange.to, 'to')
    });
  }
}

const periodValidator = ({
  min = new Date(0),
  max = new Date(3000, 0, 0, 0, 0, 0)
}: { min?: Date; max?: Date } = {}): ValidatorFn => {
  return control => {
    if (!control.value) {
      return null;
    }

    if (control.value.match(/[0-9]{4}-[0-9]{2}-[0-9]{2}(T[0-9]{2}:[0-9]{2}:[0-9]{2})?/g)) {
      const date = new Date(new Date(control.value).setHours(0, 0, 0, 0));
      return !isNaN(date.getTime()) && date >= min && date <= max ? null : { period: 'notValid' };
    }

    const matches = control.value.match(RELATIVE_TIME_REGEX) ?? [];
    if (matches[0] === control.value) {
      return null;
    }

    return { period: 'notValid' };
  };
};
