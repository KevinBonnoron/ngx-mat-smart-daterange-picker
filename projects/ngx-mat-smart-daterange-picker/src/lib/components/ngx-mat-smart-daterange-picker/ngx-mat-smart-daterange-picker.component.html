<div>
  <ng-template #popoverContent>
    <div fxFlex="row" style="width: 1100px">
      <div fxFlex="20" fxFlex="1 1 80%">
        <ngx-daterange-picker [value]="value" [min]="min" [max]="max" [maxRange]="maxRange"
          (valueChange)="dateRangePickerChange($event)">
        </ngx-daterange-picker>
      </div>
      <div fxFlex="30" fxFlexOffset="5" fxFlex="column">
        <ngx-absolute-timerange-picker [value]="value" [min]="min" [max]="max"
          (valueChanged)="absoluteTimeRangeChange($event)">
        </ngx-absolute-timerange-picker>
        <ngx-recently-used-time-range [dateRanges]="recentlyUsedTimeRanges"
          (selectionChange)="recentlyUsedTimeRangeChange($event)"></ngx-recently-used-time-range>
      </div>
      <div fxFlex="1 0 auto" fxFlexOffset="5" class="left-separator">
        <ngx-relative-timerange-picker [relativeTimeRanges]="relativeTimeRanges"
          (selectionChange)="relativeTimeRangeChange($event)">
        </ngx-relative-timerange-picker>
      </div>
    </div>
  </ng-template>

  <ngx-mat-popover [popoverContent]="popoverContent" [xPosition]="'after'" [yPosition]="'below'">
    <button mat-raised-button mat-button [disabled]="disabled">
      <mat-icon>schedule</mat-icon>&nbsp;{{ timeRangeLabel }}
    </button>
  </ngx-mat-popover>
</div>