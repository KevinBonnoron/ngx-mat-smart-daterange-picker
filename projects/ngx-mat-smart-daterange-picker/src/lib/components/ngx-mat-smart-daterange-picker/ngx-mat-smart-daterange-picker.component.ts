import { coerceBooleanProperty } from '@angular/cdk/coercion';
import {
  Component,
  ElementRef,
  HostBinding,
  Inject,
  Input,
  LOCALE_ID,
  OnDestroy,
  OnInit,
  Optional,
  Self,
  ViewChild
} from '@angular/core';
import { ControlValueAccessor, NgControl } from '@angular/forms';
import { MatFormFieldControl } from '@angular/material/form-field';
import { NgxMatPopoverComponent } from 'ngx-mat-popover';
import { Subject } from 'rxjs';
import { DateRange, Queue, RelativeTimeRange, UniqueQueue } from '../../types';
import { formatDateRange, isDateRangeEquals } from '../../utils';

@Component({
  selector: 'ngx-mat-smart-daterange-picker',
  templateUrl: './ngx-mat-smart-daterange-picker.component.html',
  providers: [
    {
      provide: MatFormFieldControl,
      useExisting: NgxMatSmartDateRangePickerComponent
    }
  ],
  exportAs: 'ngxSmartDateRangePicker'
})
export class NgxMatSmartDateRangePickerComponent
  implements OnInit, OnDestroy, MatFormFieldControl<DateRange>, ControlValueAccessor
{
  static nextId = 0;

  @ViewChild(NgxMatPopoverComponent, { static: true })
  readonly ngxMatPopover!: NgxMatPopoverComponent;

  @Input()
  min?: Date;

  @Input()
  max?: Date;

  @Input()
  maxRange = 0;

  @Input()
  recentlyUsedTimeRangeCount = 5;

  @Input()
  relativeTimeRanges: RelativeTimeRange[] | undefined = undefined;

  @HostBinding()
  id = `mat-smart-datetime-picker-${NgxMatSmartDateRangePickerComponent.nextId++}`;
  stateChanges = new Subject<void>();
  private _value!: DateRange;
  private _placeholder: string = '';
  private _focused: boolean = false;
  private _required: boolean = false;
  private _disabled = false;
  private onChange: (_: any) => void = () => {};
  private onTouched = () => {};
  controlType = 'mat-smart-datetime-picker';
  autofilled?: boolean | undefined;
  userAriaDescribedBy?: string | undefined;
  recentlyUsedTimeRanges!: Queue<DateRange>;

  constructor(
    @Optional() @Self() public readonly ngControl: NgControl,
    private readonly _elementRef: ElementRef,
    @Inject(LOCALE_ID) private readonly locale: string
  ) {
    if (this.ngControl != null) {
      // Setting the value accessor directly (instead of using
      // the providers) to avoid running into a circular import.
      this.ngControl.valueAccessor = this;
    }
  }

  ngOnInit(): void {
    this.recentlyUsedTimeRanges = new UniqueQueue<DateRange>(this.recentlyUsedTimeRangeCount, isDateRangeEquals);
    if (this.value) {
      this.recentlyUsedTimeRanges.push(this.value);
    }
  }

  ngOnDestroy(): void {
    this.stateChanges.complete();
  }

  dateRangePickerChange(dateRange: DateRange): void {
    this.value = dateRange;
  }

  absoluteTimeRangeChange(dateRange: DateRange): void {
    this.ngxMatPopover.close();
    this.value = dateRange;
  }

  recentlyUsedTimeRangeChange(dateRange: DateRange): void {
    this.ngxMatPopover.close();
    this.value = dateRange;
  }

  relativeTimeRangeChange(dateRange: DateRange): void {
    this.ngxMatPopover.close();
    this.value = dateRange;
  }

  writeValue(value: DateRange): void {
    this._value = value;
  }

  registerOnChange(onChange: any): void {
    this.onChange = onChange;
  }

  registerOnTouched(onTouched: any): void {
    this.onTouched = onTouched;
  }

  setDescribedByIds(ids: string[]): void {
    const controlElement = this._elementRef.nativeElement.querySelector('button')!;
    controlElement.setAttribute('aria-describedby', ids.join(' '));
  }

  onContainerClick(event: MouseEvent): void {
    // Nothing to do
  }

  onFocusIn(event: FocusEvent) {
    if (!this._focused) {
      this._focused = true;
      this.stateChanges.next();
    }
  }

  onFocusOut(event: FocusEvent) {
    if (!this._elementRef.nativeElement.contains(event.relatedTarget as Element)) {
      this._focused = false;
      this.onTouched();
      this.stateChanges.next();
    }
  }

  get timeRangeLabel() {
    if (this.value) {
      return formatDateRange(this.value, this.locale);
    }

    return '';
  }

  @Input()
  get value(): DateRange {
    return this._value;
  }

  set value(value: DateRange) {
    if (value?.from && value?.to && JSON.stringify(value) !== JSON.stringify(this._value)) {
      this.recentlyUsedTimeRanges?.push(value);

      this._value = value;
      this.stateChanges.next();
      this.onChange(value);
    }
  }

  @Input()
  get placeholder() {
    return this._placeholder;
  }

  set placeholder(placeholder) {
    this._placeholder = placeholder;
    this.stateChanges.next();
  }

  get empty(): boolean {
    return this.value === null;
  }

  @HostBinding('class.floating')
  get shouldLabelFloat(): boolean {
    return this._focused || !this.empty;
  }

  @Input()
  get required() {
    return this._required;
  }

  set required(req) {
    this._required = coerceBooleanProperty(req);
    this.stateChanges.next();
  }

  @Input()
  get disabled(): boolean {
    return this._disabled;
  }

  set disabled(value: boolean) {
    this._disabled = coerceBooleanProperty(value);
    // TODO
    // this._disabled ? this.parts.disable() : this.parts.enable();
    this.stateChanges.next();
  }

  get errorState(): boolean {
    return this.ngControl?.invalid ?? false;
  }

  get focused(): boolean {
    return this._focused;
  }
}
