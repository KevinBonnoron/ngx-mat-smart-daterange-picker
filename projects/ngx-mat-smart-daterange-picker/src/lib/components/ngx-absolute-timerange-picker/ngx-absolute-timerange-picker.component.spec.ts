import { ReactiveFormsModule } from '@angular/forms';
import { MockBuilder, MockRenderFactory, ngMocks } from 'ng-mocks';
import { NgxSmartDateRangePickerModule } from '../../ngx-mat-smart-daterange-picker.module';
import { NgxAbsoluteTimeRangePickerComponent } from './ngx-absolute-timerange-picker.component';

describe('NgxAbsoluteTimeRangePickerComponent', () => {
  const factory = MockRenderFactory(NgxAbsoluteTimeRangePickerComponent);

  ngMocks.faster();

  beforeAll(() =>
    MockBuilder(NgxAbsoluteTimeRangePickerComponent, NgxSmartDateRangePickerModule).keep(ReactiveFormsModule)
  );
  beforeAll(() => factory.configureTestBed());

  it('should create', () => {
    const fixture = factory();
    expect(fixture).toBeDefined();
  });

  it('should have the expected render', () => {
    const fixture = factory();
    expect(fixture).toMatchSnapshot();
  });
});
