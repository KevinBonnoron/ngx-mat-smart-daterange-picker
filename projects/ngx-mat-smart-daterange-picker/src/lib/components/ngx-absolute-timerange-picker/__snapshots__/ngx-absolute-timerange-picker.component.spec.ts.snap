// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NgxAbsoluteTimeRangePickerComponent should have the expected render 1`] = `
<mock-render
  __ngContext__={[Function LRootView]}
>
  <ngx-absolute-timerange-picker>
    <div>
      <h3>
      </h3>
      <form
        autocomplete="off"
        class="ng-untouched ng-pristine ng-invalid"
        novalidate=""
      >
        <div
          fxlayout="column"
        >
          <mat-form-field>
            <mat-label>
            </mat-label>
            <input
              class="ng-untouched ng-pristine ng-invalid"
              formcontrolname="from"
              matinput=""
            />
          </mat-form-field>
          <mat-form-field>
            <mat-label>
            </mat-label>
            <input
              class="ng-untouched ng-pristine ng-invalid"
              formcontrolname="to"
              matinput=""
            />
          </mat-form-field>
          <button
            color="primary"
            mat-raised-button=""
          >
          </button>
        </div>
      </form>
    </div>
  </ngx-absolute-timerange-picker>
</mock-render>
`;
